cirq-core/cirq/optimizers/transformer_primitives.py

```diff
@@ -129,6 +130,118 @@ def map_operations_and_unroll(
+        if new_op is not None and not qubit_set.issuperset(new_op.qubits):
+            raise ValueError(
+                f"Merged operation {new_op} must act on a subset of qubits of "
+                f"original operations {op1} and {op2}"
+            )
+        return new_op
+
+    ret_circuit = circuits.Circuit()
+    for current_moment in circuit:
+        new_moment = ops.Moment()
+        for op in current_moment:
+            op_qs = set(op.qubits)
+            idx = ret_circuit.prev_moment_operating_on(tuple(op_qs))
+            if idx is not None and op_qs.issubset(ret_circuit[idx][op_qs].operations[0].qubits):
+                # Case-1: Try to merge op with the larger operation on the left.
+                left_op = ret_circuit[idx][op_qs].operations[0]
+                new_op = apply_merge_func(left_op, op)
+                if new_op is not None:
+                    ret_circuit.batch_replace([(idx, left_op, new_op)])
+                else:
+                    new_moment = new_moment.with_operation(op)
+                continue
+
+            while idx is not None and len(op_qs) > 0:
+                # Case-2: left_ops will merge right into `op` whenever possible.
+                for left_op in ret_circuit[idx][op_qs].operations:
+                    is_merged = False
+                    if op_qs.issuperset(left_op.qubits):
+                        # Try to merge left_op into op
+                        new_op = apply_merge_func(left_op, op)
+                        if new_op is not None:
+                            ret_circuit.batch_remove([(idx, left_op)])
+                            op, is_merged = new_op, True
+                    if not is_merged:
+                        op_qs -= frozenset(left_op.qubits)
+                idx = ret_circuit.prev_moment_operating_on(tuple(op_qs))
+            new_moment = new_moment.with_operation(op)
+        ret_circuit += new_moment
+    return _to_target_circuit_type(ret_circuit, circuit)
+
+
+def merge_moments(
+    circuit: CIRCUIT_TYPE,
+    merge_func: Callable[[ops.Moment, ops.Moment], Optional[ops.Moment]],
+) -> CIRCUIT_TYPE:
+    """Merges adjacent moments, one by one from left to right, by calling `merge_func(m1, m2)`.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        merge_func: Callable to determine whether two adjacent moments in the circuit should be
+            merged. If the moments can be merged, the callable should return the merged moment,
+            else None.
+
+    Returns:
+        Copy of input circuit with merged moments.
+    """
+    if not circuit:
+        return circuit
+    merged_moments: List[ops.Moment] = [circuit[0]]
+    for current_moment in circuit[1:]:
+        merged_moment = merge_func(merged_moments[-1], current_moment)
+        if not merged_moment:
+            merged_moments.append(current_moment)
+        else:
+            merged_moments[-1] = merged_moment
```

Comment:
This feels like it might be able to be converted to a list comprehension of some kind. Can it be ?

Reply:
I couldn't come up with a straightforward way to do it. LMK if you have suggestions.